interpreter.go

```diff
@@ -267,150 +186,120 @@ func makeCallStack(limit int) callStack {
-	switch curr := curr.(type) {
-	// case *valueExtendedObject:
-	// TODO(dcunnin): this
-
-	case *valueSimpleObject:
-		if *counter >= startFrom {
-			if _, ok := curr.fields[f]; ok {
-				return curr
-			}
-		}
-		*counter++
 
-		// case *valueComprehensionObject:
-		/*
-			if *counter >= startFrom {
-				// TODO(dcunnin): this
-			}
-			*counter++
-		*/
-	}
-	return nil
+	return result
 }
 
-func (i *interpreter) objectIndex(loc *LocationRange, obj value, f string, offset int) (astNode, error) {
-	var foundAt int
-	self := obj
-	found := i.findObject(f, obj, offset, &foundAt)
-	if found == nil {
-		return nil, makeRuntimeError(loc, fmt.Sprintf("Field does not exist: %s", f))
-	}
-	switch found := found.(type) {
-	case *valueSimpleObject:
-		field := found.fields[f]
-		i.stack.newCall(loc, found, self, foundAt, found.upValues)
-		return field.body, nil
-	// case *valueComprehensionObject:
-	/*
-		// TODO(dcunnin): this
-	*/
-	default:
-		return nil, fmt.Errorf("Internal error: findObject returned unrecognized type: %s", reflect.TypeOf(found))
-	}
+func (i *interpreter) getCurrentEnv(ast astNode) environment {
+	return makeEnvironment(
+		i.capture(ast.FreeVariables()),
+		i.stack.getSelfBinding(),
+	)
 }
 
-func (i *interpreter) evalAdd(leftVal, rightVal value) (value, error) {
-	// TODO(sbarzowski) More types and more graceful error handling
-	switch leftVal := leftVal.(type) {
-	case *valueNumber:
-		left := leftVal.value
-		right := rightVal.(*valueNumber).value
-		return makeValueNumber(left + right), nil
-	case *valueString:
-		left := leftVal.value
-		right := rightVal.(*valueString).value
-		return makeValueString(left + right), nil
-	default:
-		panic(fmt.Sprintf("INTERNAL ERROR: Unrecognised value type: %T", leftVal))
+func (i *interpreter) evaluate(a astNode, context *TraceContext) (value, error) {
+	// TODO(dcunnin): All the other cases...
+
+	e := &evaluator{
+		trace: &TraceElement{
+			loc:     a.Loc(),
+			context: context,
+		},
+		i: i,
 	}
-}
 
-func (i *interpreter) evaluate(a astNode) (value, error) {
-	// TODO(dcunnin): All the other cases...
 	switch ast := a.(type) {
 	case *astArray:
-		self, offset := i.stack.getSelfBinding()
-		var elements []*thunk
+		sb := i.stack.getSelfBinding()
+		var elements []potentialValue
 		for _, el := range ast.elements {
-			elThunk := makeThunk(i.idArrayElement, self, offset, el)
-			elThunk.upValues = i.capture(el.FreeVariables())
+			env := makeEnvironment(i.capture(el.FreeVariables()), sb)
+			elThunk := makeThunk(i.idArrayElement, env, el)
 			elements = append(elements, elThunk)
 		}
-		return &valueArray{elements}, nil
+		return makeValueArray(elements), nil
 
 	case *astBinary:
-		// TODO(dcunnin): Assume it's + for now
-		leftVal, err := i.evaluate(ast.left)
+		// Some binary operators are lazy, so thunks are needed in general
+		env := i.getCurrentEnv(ast)
+		left := makeThunk("x", env, ast.left)
```

Comment:
"x" & "y" are thunk names? what are the names for? these don't seem very descriptive/helpful